File: server/routes/llm-search.js
const express = require("express");
const fs = require("fs");
const path = require("path");
const { db } = require("../db");
const TfIdfSearchEngine = require("../search-engine");
const router = express.Router();

const engine = new TfIdfSearchEngine();
let indexed = false;

// ─── LLM 설정 파일 관리 ───
const CONFIG_PATH = path.join(__dirname, "..", "llm-config.json");

function loadConfig() {
  try {
    if (fs.existsSync(CONFIG_PATH)) {
      return JSON.parse(fs.readFileSync(CONFIG_PATH, "utf-8"));
    }
  } catch (e) {
    console.error("LLM config load error:", e.message);
  }
  return null;
}

function saveConfig(config) {
  fs.writeFileSync(CONFIG_PATH, JSON.stringify(config, null, 2), "utf-8");
}

function getConfig() {
  const saved = loadConfig();
  if (saved) return saved;
  // 환경변수 폴백
  if (process.env.LLM_BASE_URL) {
    return {
      baseUrl: process.env.LLM_BASE_URL,
      apiKey: process.env.LLM_API_KEY || "",
      model: process.env.LLM_MODEL || "gpt-120b",
    };
  }
  return null;
}

function ensureIndex() {
  const experiments = db
    .prepare(
      `
    SELECT e.*,
      p.project_purpose, p.iacpj_ta_goa, p.iacpj_cur_stt,
      p.iacpj_tech_n, p.iacpj_mod_n as project_module,
      p.iacpj_tgt_n, p.iacpj_level,
      p.ia_tgt_htr_n, p.iacpj_nud_n,
      p.iacpj_itf_uno, p.iacpj_bgn_dy, p.iacpj_ch_n, p.ia_ta_grd_n
    FROM experiments e
    LEFT JOIN projects p ON e.iacpj_nm = p.iacpj_nm
  `,
    )
    .all();
  engine.buildIndex(experiments);
  indexed = true;
}

// 실험 데이터를 LLM 컨텍스트용 텍스트로 변환
function experimentToContext(result, idx) {
  const exp = result.experiment;
  const proj = result.project;
  const splits = result.splits || [];

  let text = `[실험 ${idx + 1}] ${exp.eval_item || "N/A"}
  - 과제: ${exp.iacpj_nm || "N/A"}
  - 모듈: ${exp.module || "N/A"}
  - Plan ID: ${exp.plan_id || "미배정"}
  - 평가공정: ${exp.eval_process || "N/A"}
  - 평가분류: ${exp.eval_category || "N/A"}
  - 요청자: ${exp.requester || "N/A"} / 팀: ${exp.team || "N/A"}
  - LOT: ${exp.lot_code || "N/A"}
  - 상태: ${exp.status || "N/A"}`;

  if (proj) {
    text += `
  - 과제 목적: ${proj.project_purpose || "N/A"}
  - 과제 목표: ${proj.iacpj_ta_goa || "N/A"}
  - 현재 상태: ${proj.iacpj_cur_stt || "N/A"}
  - 대상 기술: ${proj.iacpj_tech_n || "N/A"}
  - 분류: ${proj.iacpj_tgt_n || "N/A"}`;
  }

  if (splits.length > 0) {
    text += `\n  - Split 조건 (${splits.length}건):`;
    for (const s of splits.slice(0, 5)) {
      text += `\n    * ${s.oper_nm || s.oper_id || "공정"}: ${s.eps_lot_gbn_cd || "base"} | ${s.work_cond_desc || "-"} | 장비: ${s.eqp_id || "-"}`;
    }
    if (splits.length > 5) text += `\n    ... 외 ${splits.length - 5}건`;
  }

  return text;
}

// LLM API 호출 (OpenAI 호환)
async function callLLM(config, systemPrompt, userMessage, conversationHistory = []) {
  const messages = [
    { role: "system", content: systemPrompt },
    ...conversationHistory,
    { role: "user", content: userMessage },
  ];

  const res = await fetch(`${config.baseUrl}/chat/completions`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      ...(config.apiKey ? { Authorization: `Bearer ${config.apiKey}` } : {}),
    },
    body: JSON.stringify({
      model: config.model,
      messages,
      temperature: 0.3,
      max_tokens: 1500,
    }),
  });

  if (!res.ok) {
    const errText = await res.text();
    throw new Error(`LLM API error (${res.status}): ${errText}`);
  }

  const data = await res.json();
  return data.choices?.[0]?.message?.content || "응답을 생성하지 못했습니다.";
}

const SYSTEM_PROMPT = `당신은 반도체 실험 데이터베이스 검색 도우미입니다.
사용자가 실험을 찾거나 분석을 요청하면, 제공된 실험 데이터를 기반으로 정확하고 간결하게 답변합니다.

규칙:
1. 제공된 실험 데이터만을 근거로 답변하세요. 데이터에 없는 내용을 추측하지 마세요.
2. 실험 간 비교, 유사 실험 추천, 조건 분석 등을 수행할 수 있습니다.
3. 답변은 한국어로 작성하세요.
4. 구조화된 형태(번호, 불렛)로 요약해주세요.
5. 사용자가 특정 실험에 대해 질문하면 해당 실험의 상세 정보를 안내하세요.`;

// ─── 설정 조회 ───
router.get("/config", (req, res) => {
  const config = getConfig();
  if (!config) {
    return res.json({ configured: false });
  }
  res.json({
    configured: true,
    baseUrl: config.baseUrl,
    model: config.model,
    // apiKey는 마스킹해서 전달
    apiKeyHint: config.apiKey
      ? config.apiKey.slice(0, 4) + "****" + config.apiKey.slice(-4)
      : "",
  });
});

// ─── 설정 등록/수정 ───
router.post("/config", (req, res) => {
  const { baseUrl, apiKey, model } = req.body;
  if (!baseUrl || !model) {
    return res.status(400).json({ error: "baseUrl과 model은 필수입니다." });
  }
  const config = { baseUrl: baseUrl.replace(/\/+$/, ""), apiKey: apiKey || "", model };
  saveConfig(config);
  res.json({
    message: "LLM 설정이 저장되었습니다.",
    configured: true,
    baseUrl: config.baseUrl,
    model: config.model,
    apiKeyHint: config.apiKey
      ? config.apiKey.slice(0, 4) + "****" + config.apiKey.slice(-4)
      : "",
  });
});

// ─── 설정 삭제 ───
router.delete("/config", (req, res) => {
  try {
    if (fs.existsSync(CONFIG_PATH)) fs.unlinkSync(CONFIG_PATH);
  } catch (_) {}
  res.json({ message: "LLM 설정이 삭제되었습니다.", configured: false });
});

// ─── 연결 테스트 ───
router.post("/config/test", async (req, res) => {
  const { baseUrl, apiKey, model } = req.body;
  if (!baseUrl || !model) {
    return res.status(400).json({ error: "baseUrl과 model은 필수입니다." });
  }
  try {
    const testConfig = { baseUrl: baseUrl.replace(/\/+$/, ""), apiKey: apiKey || "", model };
    const answer = await callLLM(testConfig, "You are a test assistant.", "Say 'OK' in Korean.", []);
    res.json({ success: true, response: answer });
  } catch (err) {
    res.json({ success: false, error: err.message });
  }
});

// ─── 메인 검색 엔드포인트 ───
router.post("/", async (req, res) => {
  const { query, conversationHistory = [] } = req.body;
  if (!query) return res.status(400).json({ error: "query is required" });

  if (!indexed) ensureIndex();

  // 후속 질문 시 대화 이력의 이전 사용자 쿼리를 합쳐 검색 범위 확장
  const prevUserQueries = conversationHistory
    .filter((m) => m.role === "user")
    .slice(-3)
    .map((m) => m.content)
    .join(" ");
  const expandedQuery = prevUserQueries ? `${query} ${prevUserQueries}` : query;

  // TF-IDF 검색: 현재 쿼리 + 확장 쿼리 병합
  const primaryResults = engine.search(query, 15);
  const expandedResults = prevUserQueries ? engine.search(expandedQuery, 15) : [];

  // 중복 제거 후 병합 (현재 쿼리 결과 우선)
  const seenIds = new Set(primaryResults.map((r) => r.document.id));
  const mergedResults = [
    ...primaryResults,
    ...expandedResults.filter((r) => !seenIds.has(r.document.id)),
  ].slice(0, 15);

  const enriched = mergedResults.map((r) => {
    const splits = db
      .prepare("SELECT * FROM split_tables WHERE plan_id = ?")
      .all(r.document.plan_id);
    const project = db
      .prepare("SELECT * FROM projects WHERE iacpj_nm = ?")
      .get(r.document.iacpj_nm);
    return {
      score: Math.round(r.score * 1000) / 1000,
      experiment: r.document,
      project,
      splits,
    };
  });

  const queryTokens = engine.tokenize(query);
  const filtered = enriched.filter((r) => {
    const docText = engine._docToText(r.experiment).toLowerCase();
    return queryTokens.every((token) => docText.includes(token));
  });
  const finalResults = filtered.length > 0 ? filtered : enriched;

  // LLM 설정 확인
  const config = getConfig();
  if (!config) {
    return res.json({
      answer: `LLM이 설정되지 않았습니다. 상단에서 모델을 등록해주세요.\n\n키워드 검색 결과: "${query}"에 대해 ${finalResults.length}건의 실험이 검색되었습니다.`,
      results: finalResults,
      resultCount: finalResults.length,
      llmError: true,
    });
  }

  try {
    const contextText = finalResults.length > 0
      ? finalResults.map((r, i) => experimentToContext(r, i)).join("\n\n")
      : "관련 실험 데이터를 찾지 못했습니다.";

    const userMessage = `[검색된 실험 데이터]\n${contextText}\n\n[사용자 질문]\n${query}\n\n위 실험 데이터를 바탕으로 사용자의 질문에 답변해주세요.`;

    const llmResponse = await callLLM(config, SYSTEM_PROMPT, userMessage, conversationHistory);

    res.json({
      answer: llmResponse,
      results: finalResults,
      resultCount: finalResults.length,
    });
  } catch (err) {
    console.error("LLM Search Error:", err);
    res.json({
      answer: `LLM 호출에 실패하여 키워드 검색 결과만 표시합니다.\n(${err.message})\n\n검색어 "${query}"에 대해 ${finalResults.length}건의 실험이 검색되었습니다.`,
      results: finalResults,
      resultCount: finalResults.length,
      llmError: true,
    });
  }
});

function invalidateIndex() {
  indexed = false;
}

module.exports = router;
module.exports.invalidateIndex = invalidateIndex;

File: server_py/routes/llm_search.py
"""
/api/llm-search — LLM-assisted search with TF-IDF pre-filtering.
"""
import re
import json
import os
import sqlite3
from fastapi import APIRouter, Depends, HTTPException
from ..database import get_db, dict_row, dict_rows
from ..search_engine import TfIdfSearchEngine

router = APIRouter(prefix="/api/llm-search", tags=["llm-search"])

engine = TfIdfSearchEngine()
_indexed = False

CONFIG_PATH = os.path.join(os.path.dirname(__file__), "..", "llm-config.json")

def _load_config():
    try:
        if os.path.exists(CONFIG_PATH):
            with open(CONFIG_PATH, "r", encoding="utf-8") as f:
                return json.load(f)
    except Exception:
        pass
    return None

def _save_config(config):
    with open(CONFIG_PATH, "w", encoding="utf-8") as f:
        json.dump(config, f, indent=2, ensure_ascii=False)

def _get_config():
    saved = _load_config()
    if saved:
        return saved
    base_url = os.environ.get("LLM_BASE_URL")
    if base_url:
        return {
            "baseUrl": base_url,
            "apiKey": os.environ.get("LLM_API_KEY", ""),
            "model": os.environ.get("LLM_MODEL", "gpt-120b"),
        }
    return None

def invalidate_index():
    global _indexed
    _indexed = False

def _ensure_index(conn):
    global _indexed
    rows = conn.execute("""
        SELECT e.*,
          p.project_purpose, p.iacpj_ta_goa, p.iacpj_cur_stt,
          p.iacpj_tech_n, p.iacpj_mod_n as project_module,
          p.iacpj_tgt_n, p.iacpj_level,
          p.ia_tgt_htr_n, p.iacpj_nud_n,
          p.iacpj_itf_uno, p.iacpj_bgn_dy, p.iacpj_ch_n, p.ia_ta_grd_n,
          p.iacpj_core_tec, p.ia_ch_or_n
        FROM experiments e
        LEFT JOIN projects p ON e.iacpj_nm = p.iacpj_nm
    """).fetchall()
    docs = []
    for exp in rows:
        d = dict(exp)
        splits = conn.execute(
            "SELECT fac_id,oper_id,oper_nm,eps_lot_gbn_cd,work_cond_desc,eqp_id,recipe_id,note FROM split_tables WHERE plan_id=?",
            (d.get("plan_id"),),
        ).fetchall()
        d["_splits"] = [dict(s) for s in splits]
        docs.append(d)
    engine.build_index(docs)
    _indexed = True

def _exp_to_context(result, idx, total):
    exp = result["experiment"]
    proj = result.get("project")
    splits = result.get("splits", [])
    text = (
        f"[후보 {idx+1}/{total}] {exp.get('eval_item','N/A')}\n"
        f"  - 과제: {exp.get('iacpj_nm','N/A')}\n"
        f"  - 모듈: {exp.get('module','N/A')}\n"
        f"  - Plan ID: {exp.get('plan_id','미배정')}\n"
        f"  - 평가공정: {exp.get('eval_process','N/A')}\n"
        f"  - 평가분류: {exp.get('eval_category','N/A')}\n"
        f"  - 요청자: {exp.get('requester','N/A')} / 팀: {exp.get('team','N/A')}\n"
        f"  - LOT: {exp.get('lot_code','N/A')}\n"
        f"  - 상태: {exp.get('status','N/A')}"
    )
    if proj:
        text += (
            f"\n  - 과제 목적: {proj.get('project_purpose','N/A')}"
            f"\n  - 과제 목표: {proj.get('iacpj_ta_goa','N/A')}"
            f"\n  - 현재 상태: {proj.get('iacpj_cur_stt','N/A')}"
            f"\n  - 대상 기술: {proj.get('iacpj_tech_n','N/A')}"
            f"\n  - 분류: {proj.get('iacpj_tgt_n','N/A')}"
        )
    if splits:
        text += f"\n  - Split 조건 ({len(splits)}건):"
        for s in splits[:5]:
            text += f"\n    * {s.get('oper_nm') or s.get('oper_id','공정')}: {s.get('eps_lot_gbn_cd','base')} | {s.get('work_cond_desc','-')} | 장비: {s.get('eqp_id','-')}"
        if len(splits) > 5:
            text += f"\n    ... 외 {len(splits)-5}건"
    return text

SYSTEM_PROMPT = """당신은 반도체 실험 데이터베이스 탐색 도우미입니다.
목표: 대화를 통해 후보 실험을 점차 1개로 좁혀나가는 것입니다.

응답 규칙:
1. 현재 후보 실험들의 핵심 차이점을 구체적으로 설명하세요.
2. 범위를 좁힐 수 있는 질문을 1~2개 제시하세요.
3. 사용자의 답변을 반영해 어떤 실험이 더 적합한지 안내하세요.
4. 후보가 1개로 좁혀지면 해당 실험의 상세 정보를 안내하고, 이 실험을 기반으로 신규 실험을 구성할지 물어보세요.
5. 후보가 많을 때는 가장 구분력 있는 기준으로 그룹핑하여 설명하세요.

금지사항:
- 데이터에 없는 내용을 추측하거나 만들어내지 마세요.
- 영어 사용 금지 (한국어 전용).
- 단순 목록 나열 금지."""

async def _call_llm(config, system_prompt, user_message, history=None):
    import httpx
    messages = [{"role":"system","content":system_prompt}]
    if history:
        messages.extend(history)
    messages.append({"role":"user","content":user_message})
    headers = {"Content-Type":"application/json"}
    if config.get("apiKey"):
        headers["Authorization"] = f"Bearer {config['apiKey']}"
    async with httpx.AsyncClient(timeout=60) as client:
        resp = await client.post(
            f"{config['baseUrl']}/chat/completions",
            headers=headers,
            json={"model":config["model"],"messages":messages,"temperature":0.3,"max_tokens":1500},
        )
        if resp.status_code != 200:
            raise Exception(f"LLM API error ({resp.status_code}): {resp.text}")
        data = resp.json()
        return data.get("choices",[{}])[0].get("message",{}).get("content","응답을 생성하지 못했습니다.")

def _parse_query(raw):
    qt = [m.group(1).strip() for m in re.finditer(r'"([^"]+)"', raw) if m.group(1).strip()]
    nq = re.sub(r'"[^"]*"'," ",raw)
    nq = re.sub(r"\s+"," ",nq).strip()
    return qt, nq

# ── Config endpoints ──
@router.get("/config")
def get_config_endpoint():
    c = _get_config()
    if not c:
        return {"configured": False}
    hint = ""
    if c.get("apiKey"):
        k = c["apiKey"]
        hint = k[:4] + "****" + k[-4:] if len(k) >= 8 else "****"
    return {"configured":True,"baseUrl":c["baseUrl"],"model":c["model"],"apiKeyHint":hint}

@router.post("/config")
def save_config_endpoint(body: dict):
    base_url = body.get("baseUrl")
    model = body.get("model")
    if not base_url or not model:
        raise HTTPException(400,"baseUrl과 model은 필수입니다.")
    c = {"baseUrl":base_url.rstrip("/"),"apiKey":body.get("apiKey",""),"model":model}
    _save_config(c)
    hint = ""
    if c["apiKey"]:
        k = c["apiKey"]
        hint = k[:4]+"****"+k[-4:] if len(k)>=8 else "****"
    return {"message":"LLM 설정이 저장되었습니다.","configured":True,"baseUrl":c["baseUrl"],"model":c["model"],"apiKeyHint":hint}

@router.delete("/config")
def delete_config_endpoint():
    try:
        if os.path.exists(CONFIG_PATH): os.remove(CONFIG_PATH)
    except: pass
    return {"message":"LLM 설정이 삭제되었습니다.","configured":False}

@router.post("/config/test")
async def test_config(body: dict):
    base_url = body.get("baseUrl")
    model = body.get("model")
    if not base_url or not model:
        raise HTTPException(400,"baseUrl과 model은 필수입니다.")
    try:
        tc = {"baseUrl":base_url.rstrip("/"),"apiKey":body.get("apiKey",""),"model":model}
        answer = await _call_llm(tc,"You are a test assistant.","Say 'OK' in Korean.",[])
        return {"success":True,"response":answer}
    except Exception as e:
        return {"success":False,"error":str(e)}

# ── Main search ──
@router.post("/")
async def llm_search(body: dict, conn: sqlite3.Connection = Depends(get_db)):
    query = body.get("query")
    history = body.get("conversationHistory", [])
    candidate_ids = body.get("candidateIds")
    if not query:
        raise HTTPException(400,"query is required")

    global _indexed
    if not _indexed:
        _ensure_index(conn)

    qt, nq = _parse_query(query)
    candidates = []

    if candidate_ids and len(candidate_ids) > 0:
        id_set = set(candidate_ids)
        all_r = engine.search(query, len(engine.documents))
        filtered = [r for r in all_r if r["document"].get("id") in id_set]
        if filtered:
            candidates = filtered[:15]
        else:
            candidates = [{"score":0.5,"document":d} for d in engine.documents if d.get("id") in id_set]
    elif qt:
        exact = [d for d in engine.documents if all(p.lower() in engine._doc_to_text(d).lower() for p in qt)]
        if nq:
            tr = engine.search(nq, len(engine.documents))
            sm = {r["document"].get("id"):r["score"] for r in tr}
            candidates = sorted([{"score":sm.get(d.get("id"),0),"document":d} for d in exact], key=lambda x:x["score"], reverse=True)[:15]
        else:
            candidates = [{"score":1,"document":d} for d in exact[:15]]
    else:
        pr = engine.search(query, 15)
        tokens = engine.tokenize(query)
        af = [r for r in pr if all(t in engine._doc_to_text(r["document"]).lower() for t in tokens)]
        candidates = af if af else pr

    enriched = []
    for r in candidates:
        doc = r["document"]
        splits = conn.execute("SELECT * FROM split_tables WHERE plan_id=?", (doc.get("plan_id"),)).fetchall()
        proj = conn.execute("SELECT * FROM projects WHERE iacpj_nm=?", (doc.get("iacpj_nm"),)).fetchone()
        enriched.append({"score":round(r["score"]*1000)/1000,"experiment":doc,"project":dict_row(proj),"splits":dict_rows(splits)})

    config = _get_config()
    if not config:
        return {
            "answer":f"LLM이 설정되지 않았습니다. 상단에서 모델을 등록해주세요.\n\n키워드 검색 결과: \"{query}\"에 대해 {len(enriched)}건의 실험이 검색되었습니다.",
            "results":enriched,"resultCount":len(enriched),"llmError":True,
        }
    try:
        total = len(enriched)
        ctx = "\n\n".join(_exp_to_context(r,i,total) for i,r in enumerate(enriched)) if total else "관련 실험 데이터를 찾지 못했습니다."
        narrowing = candidate_ids and len(candidate_ids) > 0
        if narrowing:
            um = f"[현재 후보 실험 목록 ({total}건)]\n{ctx}\n\n[사용자 메시지]\n{query}\n\n후보 실험들의 차이점을 분석하고, 사용자의 답변을 반영하여 어떤 실험이 더 적합한지 안내하며 후보를 좁혀주세요."
        else:
            um = f"[검색된 후보 실험 목록 ({total}건)]\n{ctx}\n\n[사용자 질문]\n{query}\n\n위 후보 실험들의 핵심 차이점을 설명하고, 범위를 좁힐 수 있는 질문을 제시해주세요."
        llm_resp = await _call_llm(config, SYSTEM_PROMPT, um, history)
        return {"answer":llm_resp,"results":enriched,"resultCount":len(enriched)}
    except Exception as e:
        return {
            "answer":f"LLM 호출에 실패하여 키워드 검색 결과만 표시합니다.\n({e})\n\n검색어 \"{query}\"에 대해 {len(enriched)}건의 실험이 검색되었습니다.",
            "results":enriched,"resultCount":len(enriched),"llmError":True,
        }
